Begin Philomena Query Language parser snapshot tests

MareStare opened the pull request:
Added the initial plumbing for Philomena Query Language parser tests or PhiQL for short (since no one has given a name to this language, I decided to take that chance).

The bulk of this code implements the "testing engine" for the parser. The tests are defined declaratively like this:

```ex
QueryCase.assert_phiql(%{
  compile: &Philomena.Images.Query.compile/2,
  snapshot: "#{__DIR__}/queries.json",
  contexts: %{
    user: [
      Labeled.new(:anon, nil),
      Labeled.new(:user, %{id: "{user_id}", role: "user"}),
      Labeled.new(:assistant, %{id: "{user_id}", role: "assistant"}),
      Labeled.new(:moderator, %{id: "{user_id}", role: "moderator"}),
      Labeled.new(:admin, %{id: "{user_id}", role: "admin"})
    ],
    watch: [true, false]
  },
  test_cases: [
    wildcard: [
      "*",
      "artist:*",
      "artist:mare"
    ],
    operators: [
      "safe OR pony",
      "safe AND pony",
      "safe AND (solo OR pony)"
    ],
    authenticated_user: [
      "my:faves"
    ]
  ]
})
```

This test can be invoked either via the regular `mix test` to validate that the existing test snapshot is fresh. Or via `philomena update-phiql-tests` to update the test snapshot after a change was made to the source code. The test snapshot has this format in the simplest form:

```json 
{
  "wildcard": [
    {
      "philomena": "*",
      "opensearch": {
        "match_all": {}
      }
    }
}
```

There can be an additional `contexts` key in the test case in case if it difference between different contextsts like anon user, regular user or admin/mod user.

MareStare commented on test/support/phiql_case.ex:
```diff
@@ -0,0 +1,138 @@
+defmodule Philomena.QueryCase do
```
Todo rename QueryCase to PhiqlCase

Meow commented:
De-ponify the test queries (since I assume it doesn't really matter what they are), we're aiming for eventual debranding and more neutral texts in the software in the future.

Meow commented:
I also think calling it just "query" instead of "phiql" is less confusing. In my mind the language's name has always been just "search syntax" or "query syntax".

MareStare commented:
Marking as draft until I fix the review comments

MareStare commented:
Fixed the review comments:
- Unfortunately, there is practically no TOML encoder library existent in Elixir. But anyway, TOML isn't really good at representing nested data structures. In this case Opensearch query is such a data structure
- Deponified the test data
- Renamed PhiQL to "Philomena.SearchSyntax"
- Added tests for `filter_id` and some more test scenarios
- Changed the `user_fixtures.ex` a bit during the iterations, but then I ended up not using that module because current tests don't depend on the users being in the DB. However, I left the changes I made in `user_fixtures.ex` since I believe they may be useful in the future

MareStare commented:
We may store the snapshot in YML format instead of JSON, that will reduce the size of the snapshot a bit.

@Meow WDYT, should we use YML here instead of JSON?

Meow commented on test/support/fixtures/users_fixtures.ex:
```diff
@@ -22,19 +32,37 @@ defmodule Philomena.UsersFixtures do
-    user_fixture(attrs)
-    |> Users.User.confirm_changeset()
-    |> Repo.update!()
+    user_fixture(Map.put(attrs, :confirmed, true))
```
use |> syntax here
```
attrs
|> Map.put(...)
|> user_fixture()
```

MareStare replied:
Fixed in [4f90db5](https://github.com/philomena-dev/philomena/pull/571/commits/4f90db5bbeb5061bec03836569753b47653cf785). Had to rebase to fix merge conflicts

liamwhite commented on mix.exs:
```diff
@@ -93,7 +93,22 @@ defmodule Philomena.MixProject do
+      # Automated testing
+      {
+        :assert_value,
+        "~> 0.10.4",
```
0.10.5 is released with the fix already

MareStare replied:
Cool, it was released almost immediately after I opened a PR. Updated to [8baeb95](https://github.com/philomena-dev/philomena/pull/571/commits/8baeb95336691b2bb4c6e7d8f83d6228da6eb91f)

MareStare commented on test/support/search_syntax_case.ex:
```diff
@@ -0,0 +1,175 @@
+      |> Jason.encode!(pretty: true)
+
+    # This is really dumb, but Elixir's subprocess API doesn't support
+    # passing custom payload via stdin for a command.
```
That port API is quite an asynchronous way to solve such a synchronous problem. I especially don't like the bash script wrapper solution for protecting against zombie processes described [here](https://hexdocs.pm/elixir/1.18.3/Port.html#module-zombie-operating-system-processes).

Let's keep this one using bash for the sake of keeping this a simple synchronous function call. I'll add a link to the port API in the comment instead
